pkg/resource/deploy/providers/provider.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package providers
+
+import (
+	"fmt"
+
+	"github.com/blang/semver"
+
+	"github.com/pulumi/pulumi/pkg/tokens"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+)
+
+// A ProviderRequest is a tuple of an optional semantic version and a package name. Whenever the engine receives a
+// registration for a resource that doesn't explicitly specify a provider, the engine creates a ProviderRequest for
+// that resource's provider, using the version passed to the engine as part of RegisterResource and the package derived
+// from the resource's token.
+//
+// The source evaluator (source_eval.go) is responsible for servicing provider requests. It does this by interpreting
+// these provider requests and sending resource registrations to the engine for the providers themselves. These are
+// called "default providers".
+//
+// ProviderRequest is useful as a hash key. The engine is free to instantiate any numver of provider requests, but it
+// is free to cache requests for a provider request that is equal to one that has already been serviced.
+type ProviderRequest struct {
+	version *semver.Version
```

Comment:
You might want to call out that if you are using it as a hash key, you SHOULD call `String()`

Reply:
will do.